presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/StateFactory.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.differentialentropy;
+
+import com.facebook.presto.array.ObjectBigArray;
+import com.facebook.presto.operator.aggregation.state.AbstractGroupedAccumulatorState;
+import com.facebook.presto.spi.function.AccumulatorStateFactory;
+
+import static java.util.Objects.requireNonNull;
+
+public class StateFactory
+        implements AccumulatorStateFactory<State>
+{
+    @Override
+    public State createSingleState()
+    {
+        return new SingleState();
+    }
+
+    @Override
+    public Class<? extends State> getSingleStateClass()
+    {
+        return SingleState.class;
+    }
+
+    @Override
+    public State createGroupedState()
+    {
+        return new GroupedState();
+    }
+
+    @Override
+    public Class<? extends State> getGroupedStateClass()
+    {
+        return GroupedState.class;
+    }
+
+    public static class GroupedState
+            extends AbstractGroupedAccumulatorState
+            implements State
+    {
+        private final ObjectBigArray<StateStrategy> strategies = new ObjectBigArray<>();
+        private long size;
+
+        @Override
+        public void ensureCapacity(long size)
+        {
+            strategies.ensureCapacity(size);
+        }
+
+        @Override
+        public void setStrategy(StateStrategy strategy)
+        {
+            requireNonNull(strategy, "strategy is null");
+
+            StateStrategy previous = getStrategy();
+            if (previous != null) {
+                size -= previous.estimatedInMemorySize();
+            }
+
+            strategies.set(getGroupId(), strategy);
+            size += strategy.estimatedInMemorySize();
+        }
+
+        @Override
+        public StateStrategy getStrategy()
+        {
+            return strategies.get(getGroupId());
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return size + strategies.sizeOf();
+        }
+    }
+
+    public static class SingleState
+            implements State
+    {
+        private StateStrategy strategy;
+
+        @Override
+        public void setStrategy(StateStrategy strategy)
+        {
+            requireNonNull(strategy, "strategy is null");
```

Comment:
combine with assignment: `this.strategy = requireNonNull(strategy, "strategy is null");`

Reply:
Nice! I didn't know that `requireNonNull` returns the first arg.